### ISSUE TITLE
all-in bug (probably in 2-7 NL)

### ISSUE THREAD

**Issue by ctm** [ctm/mb2-doc#451]
> tegwin: sorry, lost it. Basically bug when big blind is all in for less than small blind. Can't do any action but fold

This was from yesterday evening's BOURBON tournament (982). My guess is it was in 2-7 NL since the no-limping rule has caused a lot of trouble in the past.

I should be able to track down the hand and then add a test and fix it for good, although it may be a few days before I can work on it.

**Comment by ctm** [comment]
Ugh. This one escaped my attention and I don't think I've fixed it, but I may have. Oh well. There's enough information in the db and in the git commit log for me to figure it out.

**Comment by ctm** [comment]
This one just bit AIYAH and it was hand #122981

**Comment by ctm** [comment]
There appears to be two different bugs here.

The first bug is that if everyone else is all-in for amounts that are all less than a player has already put in as a blind, then that player should not even be asked for what to do. At that point everyone but one player is all-in. In most cases mb2 recognizes this situation, but it doesn't if the small blind covers an all-in in the big blind, which is what happened here.

However, the second bug is that no player should ever be prohibited from putting all that player's chips in, when the amount is less than the maximum raise. There is code in mb2 that is supposed to implement that, but for some reason it's not working. I haven't figured that one out just yet. I'm writing this note so that I don't overlook the first bug.

**Comment by ctm** [comment]
So, here's the trouble code:
```
        let bet_amount = self.per_round.bet_amount;
...
        let to_call = bet_amount - this_round_action;
...
        let amount = match amount {
            Any => cmp::min(to_call, player.chips()),
            N(max) => {
                if to_call <= max {
                    cmp::min(to_call, player.chips())
                } else {
                    return self.fold(seat);
                }
            }
        };

        if !call_allowed && to_call > 0 && to_call < player.chips() {
            self.player_message(player_id, PlayerMessage::NoLimping);
            return false;
        }
```
The first section says that the `bet_amount` is the bet amount of this round, but that's not actually true in a game with blinds if the player whose turn it is to pay the big blind has less than the big blind.

The second section says that the amount `to call` is whatever is the bet amount, which would be the big blind amount, minus what the player has already put in this round, which is the small blind amount. So, in AIYAH's case, he had put in a 4,000 small blind and the big blind was 8,000, so that makes `to_call` 4,000.

However, after putting in the small blind, AIYAH still had 5,500 left, so `NoLimping` gets returned. Technically, `to_call` there should have been 0, since there was no action above what AIYAH had already put in.

There are a few ways that I can change the first section, but I have to avoid pitfalls that would break the code elsewhere or would cause trouble if we had more than two blinds. Off the top of my head, if all the opponents are all-in, then `bet_amount` should be the max of the all-in players' `this_round_action`. 

I _think_ I need to do this in some other places too and once I do the "first bug" from my previous comment will go away, which means that in the exact same situation, a player would never get to this specific instance of the fix.

**Comment by ctm** [comment]
My solution was to create a function `per_round_bet_amount(&self)` that returns `self.per_round.bet_amount` if `self.n_not_all_in > 1` otherwise it returns the max `this_round_action` for all the players.

Deployed.

**Comment by ctm** [comment]
This is still happening. Hand number 136406 was an instance of it.

**Comment by ctm** [comment]
FWIW, the instance that grizz hit appears to be slightly different than what has happened before. Both blinds were all-in, and grizz was raising the default amount. It looks like the raise of the default amount itself was returning `NoLimping` because the default amount itself was being improperly computed too low.

I should look at other places where `NoLimping` was returned in that game and see if they're the same issue.

**Comment by ctm** [comment]
I've got it fixed (and tested!) locally, but won't necessarily have the fix deployed for this evening's tournament. Since this evening's tournament is Bugs, it's not an issue.

I added some test code that will make finding and fixing bugs like this (although I hope there are no more) easier, but it's still awkward code and I want to clean it up slightly before I merge it into master, which is why I'm not deploying my fix just yet.
